core/builder.go

```diff
@@ -27,24 +28,20 @@ func NewNode(ctx context.Context, cfg *BuildCfg) (*IpfsNode, error) {
 		fx.Extract(n),
 	)
 
-	go func() {
-		// Note that some services use contexts to signal shutting down, which is
-		// very suboptimal. This needs to be here until that's addressed somehow
-		<-ctx.Done()
-		err := app.Stop(context.Background())
-		if err != nil {
-			log.Error("failure on stop: ", err)
-		}
-	}()
-
 	n.IsOnline = cfg.Online
 	n.app = app
 
 	if app.Err() != nil {
 		return nil, app.Err()
 	}
 
+	// bind the process to the context so we stop when the context is
+	// canceled
+	goprocessctx.CloseAfterContext(n.Process, ctx)
```

Comment:
We can also go back to just calling `app.Stop()` when the context is canceled if we don't want to introduce any goprocess stuff here. However, I'm not sure if `app.Stop()` is thread-safe (although I assume it is).